pgcharts.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "PGCHARTS" "1" "June 2015" "ff" ""
.
.SH "NAME"
\fBpgcharts\fR \- PostgreSQL data loader
.
.SH "SYNOPSIS"
\fBpgcharts\fR [\fIoptions\fR] [\fIcommand\-file\fR]\.\.\.
.
.SH "DESCRIPTION"
pgcharts is The PostgreSQL Extension Installer server\.
.
.SH "OPTIONS"
.
.TP
\fB\-h\fR, \fB\-\-help\fR
Show command usage summary and exit\.
.
.TP
\fB\-V\fR, \fB\-\-version\fR
Show pgcharts version string and exit\.
.
.TP
\fB\-c\fR, \fB\-\-config\fR
Use the given configuration file (default to "~/\.pgcharts\.ini")\.
.
.SH "COMMANDS"
The pgcharts binary allows running and controling the pgcharts embedded web server\.
.
.SS "CONFIGURATION CONTROL"
While it\'s possible to ship a configuration file or to prepare it by hand, the following commands allow to control the setup from the command line\.
.
.IP "\(bu" 4
\fBconfig [ name ] [ value ]\fR
.
.IP
Without arguments, print the whole configuration file content\. When given a variable \fIname\fR, print its current value\. When given both a \fIname\fR and a \fIvalue\fR, set the configuration variable to the given value\.
.
.IP "\(bu" 4
\fBconfig get <name>\fR
.
.IP
Print the current value of the configuration variable \fIname\fR\.
.
.IP "\(bu" 4
\fBconfig set <name> <value>\fR
.
.IP
Set the variable \fIname\fR to the given \fIvalue\fR\.
.
.IP "" 0
.
.SS "SERVER CONTROL"
The PostgreSQL Extension Installer comes with a PostgreSQL plugin that downloads static files: that part doesn\'t need any server at all\. This server is meant to be used by maintainers of a set of extension archives, when they want to ease the maintenance and setup of the building\.
.
.IP "\(bu" 4
\fBstart\fR
.
.IP
Start the embedded pgcharts HTTP server on the port it\'s been setup to listen to, which defaults to 8042\. The
.
.IP "\(bu" 4
\fBstop\fR
.
.IP
Stops the server\.
.
.IP "\(bu" 4
\fBstatus\fR
.
.IP
Print the result of querying the HTTP status API against the (hopefully) running server\.
.
.IP "\(bu" 4
\fBpid\fR
.
.IP
Print the registered pid of the server process\. This information might be stale in case of unexpected termination of the server\.
.
.IP "\(bu" 4
\fBsetup <dburi>\fR
.
.IP
Connects to the PostgreSQL database specified with the \fIdburi\fR parameter and install the database model there\.
.
.IP "" 0
.
.SS "REGISTERING DATABASES"
Once pgcharts is properly setup (see the \fBsetup\fR command above) then it\'s necessary to add databases against which you want to run queries and draw charts\.
.
.IP "\(bu" 4
\fBregister <dburi>\fR
.
.IP
Register given \fIdburi\fR\.
.
.IP "" 0
.
.SH "DATABASE URI"
The \fIdburi\fR connection string is expected to be given as a \fIConnection URI\fR as documented in the PostgreSQL documentation at http://www\.postgresql\.org/docs/9\.3/static/libpq\-connect\.html#LIBPQ\-CONNSTRING\.
.
.IP "" 4
.
.nf

postgresql://[user[:password]@][netloc][:port][/dbname][?sslmode=\.\.\.]
.
.fi
.
.IP "" 0
.
.P
Where:
.
.IP "\(bu" 4
\fIuser\fR
.
.IP
Can contain any character, including colon (\fB:\fR) which must then be doubled (\fB::\fR) and at\-sign (\fB@\fR) which must then be doubled (\fB@@\fR)\.
.
.IP
When omitted, the \fIuser\fR name defaults to the value of the \fBPGUSER\fR environment variable, and if it is unset, the value of the \fBUSER\fR environment variable\.
.
.IP "\(bu" 4
\fIpassword\fR
.
.IP
Can contain any character, including that at sign (\fB@\fR) which must then be doubled (\fB@@\fR)\. To leave the password empty, when the \fIuser\fR name ends with at at sign, you then have to use the syntax user:@\.
.
.IP
When omitted, the \fIpassword\fR defaults to the value of the \fBPGPASSWORD\fR environment variable if it is set, otherwise the password is left unset\.
.
.IP "\(bu" 4
\fInetloc\fR
.
.IP
Can be either a hostname in dotted notation, or an ipv4, or an Unix domain socket path\. Empty is the default network location, under a system providing \fIunix domain socket\fR that method is preferred, otherwise the \fInetloc\fR default to \fBlocalhost\fR\.
.
.IP
It\'s possible to force the \fIunix domain socket\fR path by using the syntax \fBunix:/path/to/where/the/socket/file/is\fR, so to force a non default socket path and a non default port, you would have:
.
.IP "" 4
.
.nf

postgresql://unix:/tmp:54321/dbname
.
.fi
.
.IP "" 0
.
.IP
The \fInetloc\fR defaults to the value of the \fBPGHOST\fR environment variable, and if it is unset, to either the default \fBunix\fR socket path when running on a Unix system, and \fBlocalhost\fR otherwise\.
.
.IP "\(bu" 4
\fIdbname\fR
.
.IP
Should be a proper identifier (letter followed by a mix of letters, digits and the punctuation signs comma (\fB,\fR), dash (\fB\-\fR) and underscore (\fB_\fR)\.
.
.IP
When omitted, the \fIdbname\fR defaults to the value of the environment variable \fBPGDATABASE\fR, and if that is unset, to the \fIuser\fR value as determined above\.
.
.IP "\(bu" 4
The only optional parameter supported is \fBsslmode\fR and it accepts the values \fBdisable\fR, \fBallow\fR, \fBprefer\fR and \fBrequire\fR\.
.
.IP "" 0
.
.SH "AUTHOR"
Dimitri Fontaine \fIdimitri@2ndQuadrant\.fr\fR
.
.SH "SEE ALSO"
The pgcharts source code and all documentation may be downloaded from \fIhttps://github\.com/dimitri/pgcharts/\fR\.


debian/changelog
pgcharts (1.0-1) unstable; urgency=medium

  * First release.

 -- Dimitri Fontaine <dim@tapoueh.org>  Wed, 10 Sep 2014 16:38:38 +0400


src/pgcharts.lisp
(in-package #:pgcharts)

;;;
;;; Handle the command line interface
;;;
(defvar *version-string* "1.0"
  "Version string: X.0 are development versions.")

(defvar *options*
  (list (make-option :help    "-h" "--help")
        (make-option :version "-V" "--version")
        (make-option :config  "-c" "--config" #'set-config-filename t))
  "List of allowed options for the main pgcharts command line.")

(define-condition server-error ()
  ((uri         :initarg :uri    :reader server-error-uri)
   (status-code :initarg :status :reader server-error-status-code)
   (reason      :initarg :reason :reader server-error-reason)
   (body        :initarg :body   :reader server-error-body)))

(define-condition cli-error ()
  ((mesg   :initarg :mesg   :initform nil :reader cli-error-message)
   (detail :initarg :detail :initform nil :reader cli-error-detail)
   (hint   :initarg :hint   :initform nil :reader cli-error-hint)))

(defun main (argv)
  "The main entry point for the command-line interface."
  (multiple-value-bind (args opts)
      (process-argv-options argv)

    (when (member :help opts)
      (usage args :help t)
      (uiop:quit 0))

    (when (member :version opts)
      (format t "pgcharts version ~s~%" *version-string*)
      (format t "compiled with ~a ~a~%"
                  (lisp-implementation-type)
                  (lisp-implementation-version))
      (uiop:quit 0))

    ;; don't do anything when --help or --version were given
    (let ((match (find-command-function args)))
      (if match
          (destructuring-bind (fun args) match
            (handler-case
                (handler-bind ((warning
                                #'(lambda (c)
                                    (format t "WARNING: ~a~%" c)
                                    (muffle-warning))))
                  (apply fun args))
              (cli-error (e)
                (format t
                        "ERROR: ~a~%~@[DETAIL: ~a~%~]~@[HINT: ~a~%~]"
                        (cli-error-message e)
                        (cli-error-detail e)
                        (cli-error-hint e)))
              (server-error (e)
                (format t
                        "ERROR ~d ON ~a~%~@[REASON: ~a~%~]~@[BODY: ~a~%~]"
                        (server-error-status-code e)
                        (server-error-uri e)
                        (server-error-reason e)
                        (server-error-body e)))
              (condition (c)
                (if (member :debug opts)
                    (invoke-debugger c)
                    (format t "ERROR: ~a~%" c)))))
          (usage argv)))))


;;;
;;; Actual commands
;;;
(define-command (("setup") (&optional dburi))
    "setup the pgcharts database"
  (let ((*dburi*  (or dburi *dburi*)))
    (write-config)
    (ensure-model-is-current)))

(define-command (("status") ())
    "get the status of the currently running server"
  (handler-case
      (format t "~a~%" (server-status))
    (condition (e)
      (error 'cli-error
             :mesg    "pgcharts is not running"
             :detail  (format nil "~a" e)))))

(define-command (("pid") ())
    "prints the PID of the server, if running"
  (format t "~a~%" (read-pid *pidfile*)))

(define-command (("stop") ())
    "stop the pgcharts web server"
  (kill-server))

(define-command (("start") ())
    "start the pgcharts web server"
  (check-setup)
  (let ((status (ignore-errors (server-status))))
    (unless (and status (string= "OK" status))
      (daemon:daemonize :output *logfile*
                        :error  *logfile*
                        :pidfile *pidfile*
                        :exit-parent t
                        :sigterm (lambda (sig)
                                   (declare (ignore sig))
                                   (stop-server)))
      (start-server :logs *logfile*)
      (loop :while *server-is-running*
         :do (sleep 1)))))

(define-command (("register") (dburi))
    "register a database"
  (register-db dburi))

(define-command (("config") ())
    "print current config"
  (format t "Current config file: ~s~%" *config-filename*)
  (with-open-file (config *config-filename*)
    (uiop:copy-stream-to-stream config *standard-output*)))

(define-command (("get") (key))
    "display current value for given configuration key"
  (print-config-value key))

(define-command (("set") (key &optional val))
    "edit configuration variables"
  (if val
      (setf (config-value key) val)
      (print-config-value key)))


;;;
;;; Support code for the previous commands
;;;
(defun print-config-value (key)
  "Print the value for KEY, print nothing in error cases."
  (let ((value (config-value key)))
    (when value
      (format t "~a~%" value))))

(defun server-status ()
  "Get the server status over HTTP."
  (multiple-value-bind (body status-code headers uri stream must-close reason)
      (drakma:http-request (format nil "http://localhost:~d/status" *listen-port*))
    (declare (ignore headers stream must-close))
    (if (= status-code 200)
        body
        (error 'server-error
               :uri uri :status status-code :reason reason :body body))))

(defun kill-server (&optional (sig "TERM"))
  "Send a signal to the server for it to stop"
  (when (kill-pid (read-pid *pidfile*) sig)
    (ignore-errors (delete-file *pidfile*))))

(defun register-db (dburi)
  "Register a new database server."
  (when (validate-dburi dburi)
    (destructuring-bind (name &rest rest)
        (parse-pgsql-connection-string dburi)
      (declare (ignore rest))
      (with-pgsql-connection (*dburi*)
        (make-dao 'db :dbname name :dburi dburi)))))

(defun check-setup ()
  "Signal a condition when the setup isn't ready for starting the service."
  (unless *dburi*
    (error 'cli-error
           :mesg   "dburi is unset"
           :detail "pgcharts need its own database"
           :hint   "run: pgcharts setup pgsql://user:pass@host/dbname"))

  (handler-case
      (validate-dburi *dburi*)
    (condition (e)
      (error 'cli-error
             :mesg   (format nil "~a" e)
             :detail (format nil "failed to connect to dburi ~s" *dburi*)
             :hint   (format nil "~a~%~a~%"
                             "createdb dbname; then set dburi:"
                             "run: pgcharts setup pgsql://user:pass@host/dbname"))))

  ;; If necessary, do the initial setup, or maybe upgrade the database
  (ensure-model-is-current *dburi*))
